sw/device/silicon_creator/lib/sigverify_mod_exp_otbn.c

```diff
@@ -2,11 +2,65 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
+#include "error.h"
+#include "sw/device/silicon_creator/lib/otbn_util.h"
 #include "sw/device/silicon_creator/lib/sigverify_mod_exp.h"
+#include "sw/device/silicon_creator/lib/sigverify_rsa_key.h"
+
+OTBN_DECLARE_APP_SYMBOLS(rsa);          // The OTBN rsa app.
+OTBN_DECLARE_PTR_SYMBOL(rsa, mode);     // The RSA mode of operation.
+OTBN_DECLARE_PTR_SYMBOL(rsa, n_limbs);  // The number of 256-bit limbs.
+OTBN_DECLARE_PTR_SYMBOL(rsa, in);       // The input message.
+OTBN_DECLARE_PTR_SYMBOL(rsa, out);      // The output message.
+OTBN_DECLARE_PTR_SYMBOL(rsa, modulus);  // The modulus to operate with.
+
+static const otbn_app_t kOtbnAppRsa = OTBN_APP_T_INIT(rsa);
+static const otbn_ptr_t kOtbnVarRsaMode = OTBN_PTR_T_INIT(rsa, mode);
+static const otbn_ptr_t kOtbnVarRsaNLimbs = OTBN_PTR_T_INIT(rsa, n_limbs);
+static const otbn_ptr_t kOtbnVarRsaIn = OTBN_PTR_T_INIT(rsa, in);
+static const otbn_ptr_t kOtbnVarRsaOut = OTBN_PTR_T_INIT(rsa, out);
+static const otbn_ptr_t kOtbnVarRsaModulus = OTBN_PTR_T_INIT(rsa, modulus);
+
+static const uint32_t kOtbnModeEncrypt = 1;
 
 rom_error_t sigverify_mod_exp_otbn(const sigverify_rsa_key_t *key,
                                    const sigverify_rsa_buffer_t *sig,
                                    sigverify_rsa_buffer_t *result) {
-  // TODO(#7347): Implement this.
-  return kErrorSigverifyBadExponent;
+  static const size_t size_words = kSigVerifyRsaNumBits / 32;
+  static const uint32_t n_limbs = kSigVerifyRsaNumBits / 256;
+
+  otbn_t otbn;
+  otbn_init(&otbn);
+  RETURN_IF_ERROR(otbn_load_app(&otbn, kOtbnAppRsa));
+
+  // TODO: The OTBN routines should be consistent with ibex exponent support.
+  if (key->exponent != 65537) {
+    return kErrorSigverifyBadExponent;
+  }
+
+  // Set mode so start() will jump into rsa_encrypt().
+  RETURN_IF_ERROR(otbn_copy_data_to_otbn(&otbn, sizeof(kOtbnModeEncrypt) / 4,
+                                         &kOtbnModeEncrypt, kOtbnVarRsaMode));
+
+  // Set the number of 256-bit limbs for this RSA operation.
+  RETURN_IF_ERROR(otbn_copy_data_to_otbn(&otbn, sizeof(n_limbs) / 4, &n_limbs,
+                                         kOtbnVarRsaNLimbs));
+
+  // Set the modulus.
+  RETURN_IF_ERROR(otbn_copy_data_to_otbn(&otbn, size_words, key->n.data,
+                                         kOtbnVarRsaModulus));
```

Comment:
Does the OTBN app calculate n0_inv itself? We also have this in `key->n0_inv`.

Reply:
The routine in use in this PR does, not sure about the refactored version @felixmiller is working on. Looks like `n0_inv` is 256b wide, so we sure could use it in OTBN if we wanted.

Reply:
We had a discussion on this a while ago (whether to store only 32 bit of n0inv (thats what Ibex needs) or the full 256 bit). Don't remember what we concluded, but if the decision was made to store the full 256 bit alongside the key, we should reuse it for OTBN (for the sake of performance). In `rsa_verify.s` (`rsa_verify_3072.s` respectively), the call to `compute_m0inv` (and the code of the routine) can be removed then and this should work (if n0inv is passed from IBEX).

Reply:
The overhead for storing the rest of the words for OTBN is 7 32b words per key in mask rom. I think the difference will not be as significant as RR and depending on the size of the otbn relevant code and the total number of keys we can go either way.

Reply:
@jon-flatley would you mind capturing this in an issue? Thanks!

Reply:
@alphan Filed that issue here https://github.com/lowRISC/opentitan/issues/8452

Reply:
Thanks!